course_inventory/published_date.py

```diff
@@ -75,40 +106,113 @@ def published_date_resp_parsing(self, response):
+        published_date_in_db = self.course_data_from_db[(self.course_data_from_db['published_at'].notnull())].shape
+        logger.info(f"Size of published courses from DB with published date: {published_date_in_db}")
+        course_with_pub_date_added_from_df = pd.merge(self.course_data_from_api, self.course_data_from_db,
+                                                      on='canvas_id', how='left')
+        logger.info(f"Size of course data after merging with DB data: {course_with_pub_date_added_from_df.shape}")
+        return course_with_pub_date_added_from_df
+
+    def get_published_course_date(self, course_ids, retry_list=None) -> None:
+        logger.info("Starting of get_published_course_date from API call")
+
+        with FuturesSession(max_workers=self.num_workers) as future_session:
             headers = {'Content-type': 'application/json', 'Authorization': 'Bearer ' + self.canvas_token}
-            if next_page_links is not None:
-                logger.info("Going through Next page URL set")
-                responses = []
-                for next_page_link in next_page_links:
-                    response = session.get(next_page_link, headers=headers)
-                    self.published_course_next_page_list.remove(next_page_link)
-                    responses.append(response)
+            if retry_list is not None:
+                logger.info("Going through error and pagination list")
+                responses = [future_session.get(retry_list[course]['url'], headers=headers)
+                             for course in retry_list]
             else:
                 logger.info("Initial Round of Fetching course published date")
                 responses = [
-                    session.get(f'{self.canvas_url}/api/v1/audit/course/courses/{course_id}?per_page=100',
-                                headers=headers)
+                    future_session.get(f'{self.canvas_url}/api/v1/audit/course/courses/{course_id}?per_page=100',
+                                       headers=headers)
                     for course_id in course_ids
                 ]
 
             for response in as_completed(responses):
                 self.published_date_resp_parsing(response)
 
-        if len(self.published_course_next_page_list) != 0:
-            logger.info(f"""Pagination size {len(self.published_course_next_page_list)} with published_at date items
-                         {self.published_course_next_page_list}""")
-            self.get_published_course_date(course_ids, self.published_course_next_page_list)
-
-        return self.published_course_date
+        if len(self.published_date_retry_bucket) != 0:
+            logger.info(f"""Retrying now with list size {len(self.published_date_retry_bucket)} 
+                        {self.published_date_retry_bucket}""")
+            self.get_published_course_date(course_ids, self.published_date_retry_bucket)
```

Comment:
Do we need to pass `course_ids` here? It seems like all the information that is used during extra runs is contained within the `retry_bucket`. Maybe we could make the `course_ids` parameter have a default value of `None`? That might require some additional logic.

Reply:
I don't want to change anything with this. This should be fine

Reply:
I think it seems silly to require this as a parameter if it's not used. You're right that it's working fine, but it doesn't seem like a best practice.

Reply:
It's fine for now